instructors.md

```diff
@@ -90,18 +90,395 @@ which are covered in the lesson on the shell.
+> 3. Fix the error.
+> 4. Repeat steps 2 and 3, until you have fixed all the errors.
+>
+> ~~~ {.python}
+> def another_function
+>   print "Syntax errors are annoying."
+>    print "But at least python tells us about them!"
+>   print "So they are usually not too hard to fix."
+> ~~~
+> Answer: `SyntaxError` for missing `:()` at end of first line, `IndentationError` for mismatch between second and third lines.
+> ~~~ {.python}
+> def another_function():
+>   print "Syntax errors are annoying."
+>   print "But at least python tells us about them!"
+>   print "So they are usually not too hard to fix."
+> ~~~
+
+> ## Identifying Variable Name Errors {.challenge}
+>
+> 1. Read the code below, and (without running it) try to identify what the errors are.
+> 2. Run the code, and read the error message. What type of `NameError` do you think this is? In other words, is it a string with no quotes, a misspelled variable, or a variable that should have been defined but was not?
+> 3. Fix the error.
+> 4. Repeat steps 2 and 3, until you have fixed all the errors.
+>
+> ~~~ {.python}
+> for number in range(10):
+>     # use a if the number is a multiple of 3, otherwise use b
+>     if (Number % 3) == 0:
+>         message = message + a
+>     else:
+>         message = message + "b"
+> print message
+> ~~~
+> Answer: 3 `NameError`s for `number` being misspelled, for `message` not defined, and for `a` not being in quotes.  
+> ~~~ {.python}
+> message = ""
+> for number in range(10):
+>     # use a if the number is a multiple of 3, otherwise use b
+>     if (Number % 3) == 0:
+>         message = message + "a"
+>     else:
+>         message = message + "b"
+> print message
+> abbabbabba
+
+> ## Identifying Item Errors {.challenge}
+>
+> 1. Read the code below, and (without running it) try to identify what the errors are.
+> 2. Run the code, and read the error message. What type of error is it?
+> 3. Fix the error.
+>
+> ~~~ {.python}
+> seasons = ['Spring', 'Summer', 'Fall', 'Winter']
+> print 'My favorite season is ', seasons[4]
+> ~~~
+> Answer: IndexError; the last entry is `seasons[3]`, so `seasons[4]` doesn't make sense.
+> ~~~ {.python}
+> seasons = ['Spring', 'Summer', 'Fall', 'Winter']
+> print 'My favorite season is ', seasons[-1]
+> ~~~
+
 ## [Defensive Programming](08-defensive.html)
 
+Solutions to exercises:
+> ## Pre- and post-conditions {.challenge}
+>
+> Suppose you are writing a function called `average` that calculates the average of the numbers in a list.
+> What pre-conditions and post-conditions would you write for it?
+> Compare your answer to your neighbor's:
+> can you think of a function that will pass your tests but not hers or vice versa?
+> Answer: 
+> possible pre-conditions: 
+> ~~~ {.python}
+> assert len(input)>0, 'List length must be non-zero'
+> import numbers, numpy
+> for n in input_list:
+>     assert isinstance(n, numbers.Number) and not numpy.isnan(n),  'list entries must be numbers'
```

Comment:
We haven't seen isinstance or isnan so I don't think we can expect students to come up with it in the exercises